daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala

```diff
@@ -1215,48 +1220,32 @@ private[lf] object SBuiltin {
 
   private[speedy] sealed abstract class SBUKeyBuiltin(templateId: TypeConName)
       extends OnLedgerBuiltin(1) {
-    private[this] val typ = AstUtil.TContractId(Ast.TTyCon(templateId))
+    import PartialTransaction.{KeyActive, KeyInactive, KeyMapping}
 
-    sealed trait KeyOperation {
-      // Callback from the engine returned NotFound
-      def handleKeyNotFound(machine: Machine): Boolean
-      // We saw an archive of this key.
-      def handleKeyArchived(machine: Machine, key: GlobalKey): Unit
-      def cidToSValue(cid: V.ContractId): SValue
-      def cidToSExpr(cid: V.ContractId): SExpr
-    }
-    final case object Fetch extends KeyOperation {
-      override def handleKeyNotFound(machine: Machine): Boolean =
-        machine.tryHandleSubmitMustFail()
-      override def handleKeyArchived(machine: Machine, key: GlobalKey): Unit =
-        // TODO (MK) Produce a proper error here.
-        crash(s"Could not find key $key")
-      override def cidToSValue(cid: V.ContractId) =
-        SContractId(cid)
-      override def cidToSExpr(cid: V.ContractId) =
-        SEImportValue(typ, V.ValueContractId(cid))
-    }
-    final case object Lookup extends KeyOperation {
-      override def handleKeyNotFound(machine: Machine): Boolean = {
-        machine.returnValue = SV.None
-        true
-      }
-      override def handleKeyArchived(machine: Machine, key: GlobalKey): Unit = {
-        machine.returnValue = SV.None
+    private[this] val typ = AstUtil.TContractId(Ast.TTyCon(templateId))
+    final protected def importCid(cid: V.ContractId): SEImportValue =
+      SEImportValue(typ, V.ValueContractId(cid))
+
+    // Callback from the engine returned NotFound
+    protected def handleNewKeyNotFound(machine: Machine): Boolean
```

Comment:
```suggestion
    protected def handleGlobalKeyNotFound(machine: Machine): Boolean
```
maybe a bit clearer?